spring-kafka/src/main/java/org/springframework/kafka/config/KafkaListenerEndpointRegistry.java

```diff
@@ -268,9 +268,14 @@ public void stop() {
 
 	@Override
 	public void stop(Runnable callback) {
+		Runnable callbackWrap = () -> {
+			this.running = false;
```

Comment:
The bug is confirmed. Thank you!

I think we don't need this wrapper and simple `this.running = false;` in the end of this `stop()` method should be enough.

It looks like this property was introduced here: https://github.com/spring-projects/spring-kafka/issues/1102

Sounds like `RabbitListenerEndpointRegistry` and `JmsListenerEndpointRegistry` have to be fixed the same way.
/CC @garyrussell

Reply:
According the `SmartLifecycle` stop method documentation: 
"The callback <b>must</b> be executed after the {@code SmartLifecycle} component does indeed stop".
I figure this means the isRunning status should already be changed to false before the callback is called.

Reply:
Hm. Good point!
So, how about to change the flag in the beginning of the method then instead?
Technically it doesn't matter how do we process `stop()` for those internal containers: this registry is stopped already because of its `stop()` method call.

Reply:
It feels a bit off to change the state before the actual stopping of the underlying containers.
Also, you cannot rely on the registry stop() method to be called, in fact the LifecycleProcessor will not call the stop() method but rather the stop(Runnable) variant when the lifecycle object is SmartLifecycle

Reply:
Well, I think that's right. This way any other concurrent attempts to stop this component will exit earlier because the state is already switched even if it has not finished yet.
I mean we need to have `this.running = false;` in the beginning of the method.

@garyrussell , WDYT?

Reply:
If this is the reasoning, then the other stop method method should also change the state before actually stopping the containers?

Reply:
I removed the wrapper and changed the state before calling the callback in two places. Please let me know what you think.

Reply:
I would say we can do what @artembilan suggests and move the `isRunning=false` in the simple `stop()` to the top as well, to make it consistent.